### PR TITLE
Replace a Pod if the nodeSelector changes

### PR DESCRIPTION
**Pull request by johscheuer**
Fixes: https://github.com/FoundationDB/fdb-kubernetes-operator/issues/460

### CODE REVIEW

File: controllers/cluster_controller_test.go [hunk 1]
```diff
@@ -242,124 +242,6 @@ var _ = Describe(fdbtypes.ProcessClassClusterController, func() {
 			})
 		})
 
-		Context("change the storage servers per disk", func() {
```
[the remark this answers is not in the document]
Comment (brownleej) [reply controllers/cluster_controller_test.go:245]: Definitely agree. I've also wondered whether we should refactor the tests to call reconcile directly instead of relying on the event-based reconciliation.

[the remark this answers is not in the document]
Comment (johscheuer) [reply controllers/cluster_controller_test.go:245]: This sounds interesting but I would address this in another PR/issue.

[the remark this answers is not in the document]
Comment (rbtcollins) [reply controllers/cluster_controller_test.go:245]: > 
> 
> Definitely agree. I've also wondered whether we should refactor the tests to call reconcile directly instead of relying on the event-based reconciliation.

I'm very glad to see this sentiment :) I'm curious what led to this change of heart?

**Comment by rbtcollins** [comment]
I think the code itself looks fine; and the refactoring is nice irrespective of the semantic change.

There's a number of cases where topology limitations will prevent scheduling a pod even without local storage volumes being in use. For instance, Zonal disks combined with Zonal nodepools have exactly this problem as well: change the selector, can't schedule the pod. 

So its not clear that this is really about local volumes, and see https://github.com/FoundationDB/fdb-kubernetes-operator/issues/267 where I raised the need for operators to be able to directly influence this. The downside to being conservative about this in the operator, is that if someone has just refactored their node labels, now they cannot fix things up in the operator without triggering a full migration of all the data - and this is currently very resource hungry (see https://github.com/FoundationDB/fdb-kubernetes-operator/issues/400 )

tl;dr: today, operators should be able to recover from this situation themselves, using the mechanism in #267, but they can't workaround the issue in #400 so easily. Perhaps come back to this later?

File: api/v1beta1/foundationdbcluster_types.go [hunk 2]
```diff
@@ -1848,6 +1848,16 @@ func (cluster *FoundationDBCluster) NeedsExplicitListenAddress() bool {
 	return source != nil && *source == PublicIPSourceService
 }
 
+// GetPublicIPSourceService returns the set PublicIPSource or the default PublicIPSourcePod
+func (cluster *FoundationDBCluster) GetPublicIPSourceService() PublicIPSource {
```
Comment (johscheuer) [inline api/v1beta1/foundationdbcluster_types.go:1852]:
```suggestion
func (cluster *FoundationDBCluster) GetPublicIPSource() PublicIPSource {
```

File: controllers/cluster_controller_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (brownleej) [reply controllers/cluster_controller_test.go:245]: I've been thinking about it based on the problems we've had with race conditions in the tests.

[the remark this answers is not in the document]
Comment (rbtcollins) [reply controllers/cluster_controller_test.go:245]: Ah yes.. I think that was implicit for me in the earlier discussion we had ~ 6 months ago - thank you : this helps me be able to communicate and surface such things.

**Review by brownleej — CHANGES_REQUESTED** [review]
We should rename the `GetPublicIPSourceService` method as you noted earlier. Once that change is made I think this is good to go.

**Review by brownleej — APPROVED** [review]
👍
